README.md
# acs
Audio Components Suite for Lazarus/Delphi

* Audio in/out
Windows: Standard (Wawemapper), DirectSound
Linux: Standard (/dev/dsp), ALSA, AO Live

* File in/out
Crossplatform: FLAC, Vorbis, RIFF (WAV), MPEG-3 (MP3), CD-Audio
Windows: DirectShow, Monkey's Audio (APE)

* Audio mixers
* Audio sample converters
* Audio indicators


docs/changes.txt
2019-02-27  serbod
- Removed DirectX dependancy and units from main package
* Separated Win32 and Win64 units

2016-09-23  serbod
* applied patches from cutec-chris and tcs-ulli
* cdrip, acs_mixer and demos fixes
* default driver for AudioIn
* some WaveOut and Mp3Tags fixes
* windows encoding fixed for Mixer and DX in/out 

2015-06-09  serbod
* TFileIn/Out, TAudioIn/Out subclassed and improved
+ audiodeck demo

2015-06-02  serbod
* memory bugs fixed in TWaveIn, TStdAudio, TFlacIn

2015-05-29  serbod
* AcsTags - updated

2015-05-26  serbod
* TVorbisIn - updated
* TStdAudioOut Wawemapper - updated

2015-05-23  serbod
* TWaveIn - updated

2015-05-20  serbod
* TFLACIn - updated
* TDXAudioOut - updated, prefetching fixed

2015-05-17  serbod
* TAcsCustomInput.GetData() fixed for all descendants
* TAcsAudioBuffer - separate positions for Read() and Write()
* FLAC from NewAC

2015-05-15  serbod
* ALSA and SMPEG drivers fixes for Linux

2015-05-04  serbod
+ TAcsCircularAudioBuffer
* TAcsCustomInput.Flush() renamed to Done() and all descendents
* Init() and Done() inheritance fixed

2015-05-02  serbod
! note: Total size/time not for all components. It actual only for file-based
  components. Default position in bytes, but it must be in samples.
* TAcsCustomInput.Time property replaced by PositionTime
- TAcsCustomInput.Size property moved to TAcsCustomFileIn 
- TAcsCustomOutput.Progress property moved to TAcsCustomFileIn
- TAcsCustomOutput.TotalTime property moved to TAcsCustomFileIn

2015-04-30  serbod
+ TAcsAudioBuffer - TStream-based buffer for audio samples
* TAcsThread renamed to TAcsOutThread, simplified and fixed
* TAcsCustomOutput - Byte array buffer replaced by TAcsAudioBuffer
* TAcsCustomOutput.Busy property renamed to Active
* TAcsCustomOutput.Prepare() and Done() now is non-abstract and can be inherited
* Available file and audio drivers added to list only when
  according library is loaded
* fixed inheritance bugs in some methods

2014-11-04  serbod
+ DSWrapper unit from NewAC, dswrapper.dll not needed
* TDXAudioOut and TDXAudioIn updated from NewAC

2014-10-10  serbod
+ property editors for Driver property (AudioIn, AudioOut)

2014-10-06  serbod
TACSSoundIndicator:
* range checking disabled only for pointer-based operations 
+ Pascal analogs for non-portable x86 assembler code
* fixes and optimizations

2014-10-05  serbod
* code formatted to LCL code style
* removed exceptions from design-time components
* some code rewritten to more strict FreePascal syntax
TACSMixer:
  * added error checking for external functions calls
  * fixed errors in SetDevNum

2006/07/04 17:12:45  z0m3ie
ACS 2.4 alt wiederhergestellt (unterschiedliche Sampleformate ...)

2005/12/30 13:24:27  z0m3ie
waveformat corrections, localizing

2005/12/30 12:54:42  z0m3ie
some error checks

2005/12/30 11:10:57  z0m3ie
some corrections to lazarus-linux depending things

2005/12/27 05:43:03  z0m3ie
fixed little bug with TDoneThread

2005/12/26 17:31:38  z0m3ie
fixed some problems in acs_dsfiles
fixed some problems in acs_vorbis
reworked all buffers

2005/12/20 20:11:52  z0m3ie
updated package

2005/12/18 17:01:54  z0m3ie
delphi compatibility

2005/12/04 16:54:34  z0m3ie
All classes are renamed, Style TACS... than T... to avoid conflicts with other components (eg TMixer is TACSMixer now)

2005/11/29 18:32:51  z0m3ie
bugfixes for win32 version

2005/11/28 21:57:24  z0m3ie
mostly FileOut fixes
moved PBuffer to PBuffer8
set all to dynamically Buffering

2005/11/27 16:50:33  z0m3ie
add ACS VolumeQuerry
make ACS_VolumeQuerry localizeable
some little errorfixes (buffersize for linuxdrivers was initially 0)
make TAudioIn workable

2005/10/05 20:26:36  z0m3ie
Linux changes

2005/09/18 19:28:59  z0m3ie
more progress on driver handling

2005/09/15 20:59:37  z0m3ie
start translate the documentation in the source for pasdoc

2005/09/13 21:54:11  z0m3ie
acs is localizeable now (ACS_Strings)

2005/09/13 21:53:45  z0m3ie
maked seperat driver (not complete jet)

2005/09/13 20:14:52  z0m3ie
driver handling classes (basic audio class)

2005/09/13 04:04:50  z0m3ie
First release without Components for Fileformats
only TFileIn and TFileOut are Visible

2005/09/12 22:04:52  z0m3ie
modified structure again, fileformats are now in an sperat folder.
all File In/Out classes are capsulated from TFileIn and TFileOut

2005/09/11 18:06:26  z0m3ie
first working Version

2005/09/09 21:33:42  z0m3ie
linux corrections

2005/09/08 22:19:00  z0m3ie
completed akrip based CDIn

2005/09/07 20:53:22  z0m3ie
begon to add MPEG and WMA support using DirectX

2005/09/04 17:59:37  z0m3ie
moving CDIn support to AKRip mostly
begon to add mpegin support for Win with mpg123

2005/09/01 19:55:48  z0m3ie
again Delphi corrections

2005/08/31 20:30:39  z0m3ie
Mixer Channelname work now
minior corrections for Converters

2005/08/31 14:33:16  z0m3ie
fixed delphi issue with TControlEntry

2005/08/30 22:10:55  z0m3ie
Mixer mostly completed

2005/08/28 20:31:18  z0m3ie
linux restructuring for 2.4

2005/08/28 18:35:53  z0m3ie
created Delphi package for 2.4
more Mixer stuff
updated some things for Delphi

2005/08/26 17:03:20  z0m3ie
begon to make acs resourcestring aware
more advanced tmixer for windows
restructured tmixer its better handleable now

2005/08/25 21:06:29  z0m3ie
Added TMultiMixer

2005/08/25 21:02:31  z0m3ie
TMultiMixer by Ross Levis added

2005/08/25 20:18:00  z0m3ie
Version 2.4 restructure
TCDPlayer removed (fits not in component structure)
TMP3ToWavConverter removed (fits not in component structure)

2005/08/22 20:17:01  z0m3ie
changed Headers to log
changed mail adress



src/formats/acs_mad.pas
(*
  this file is a part of audio components suite v 2.3.
  copyright (c) 2002-2005 andrei borovsky. all rights reserved.
  see the license file for more details.
  you can contact me at mail@z0m3ie.de
*)

{
$Log: acs_mad.pas,v $
Revision 1.1  2005/12/19 18:36:38  z0m3ie
*** empty log message ***

Revision 1.2  2005/12/04 16:54:34  z0m3ie
All classes are renamed, Style TAcs... than T... to avoid conflicts with other components (eg TMixer is TAcsMixer now)

Revision 1.1  2005/09/12 22:04:52  z0m3ie
modified structure again, fileformats are now in an sperat folder.
all File In/Out classes are capsulated from TFileIn and TFileOut

Revision 1.3  2005/08/25 20:18:00  z0m3ie
Version 2.4 restructure
TCDPlayer removed (fits not in component structure)
TMP3ToWavConverter removed (fits not in component structure)

Revision 1.2  2005/08/22 20:17:01  z0m3ie
changed Headers to log
changed mail adress

}

unit acs_mad;

interface

uses

  ACS_Types, Classes, SysUtils, Math, MAD;

type

  TRawPCMWaveHeader = record
    RIFF: array [0..3] of Char;
    FileSize: Integer;
    RIFFType: array [0..3] of Char;
    FmtChunkId: array [0..3] of Char;
    FmtChunkSize: Integer;
    FormatTag: Word;
    Channels: Word;
    SampleRate: Integer;
    BytesPerSecond: Integer;
    BlockAlign: Word;
    BitsPerSample: Word;
    DataChunkId: array [0..3] of Char;
    DataSize: Integer;
  end;

  TMADProgressEvent = procedure(Sender : TComponent) of object;
  TMADDoneEvent = procedure(Sender : TComponent; Success : Boolean) of object;

  TMADThread = class(TThread)
  private
    _Free : Boolean;
    Progr : Integer;
    Owner : TComponent;
    FDecoder :  mad_decoder;
    FInputStream : TStream;
    FOutputStream : TStream;
    HasFirstFrame : Boolean;
    FSR : Integer;
    FChan : Integer;
    FBitrate : Integer;
    FValid : Boolean;
    Data : PACSBuffer8;
    InputDone : Boolean;
    WaveHdr : TRawPCMWaveHeader;
    FSize : Integer;
    FMADProgress : TMADProgressEvent;
    FMADDone : TMADDoneEvent;
    WhenDone : procedure of object;
  protected
    procedure Execute; override;
  public
    constructor Create(AOwner : TComponent; InputStream, OutputStream : TStream);
    destructor Destroy; override;
  end;

implementation

   function InputFunc(CData : Pointer; Stream : p_mad_stream) : Integer; cdecl;
   var
     MT : TMADThread;
     Len : Integer;
   begin

     MT := TMADThread(CData);

     if MT.InputDone then
     begin
       Result := MAD_FLOW_STOP;
       Exit;
     end;
     MT.InputDone := True;
//     Len := MT.FInputStream.Read(Data^, MT.FInputStream.Size);
     Len := MT.FInputStream.Size;
     mad_stream_buffer(Stream, MT.Data, Len);
     if not MT.Terminated then Result := MAD_FLOW_CONTINUE
     else Result := MAD_FLOW_STOP;
   end;

   function OutputFunc(CData : Pointer; Header : p_mad_header; pcm : p_mad_pcm) : Integer; cdecl;
   var
     MT : TMADThread;
     i, framesize : Integer;
     outsamples : array[0..2303] of SmallInt;
     text : array[0..4] of Char;
     CProgr : Integer;
   begin
     MT := TMADThread(CData);
     if not MT.HasFirstFrame then
     begin
       MT.FSR := pcm.samplerate;
       MT.FChan := pcm.channels;
       MT.FBitrate := Header.bitrate;
       framesize := Ceil(144*MT.FBitrate/MT.FSR);
       MT.FSize := Round(MT.FInputStream.Size/framesize*1152)*MT.FChan*2;
       MT.FValid := True;
       text := 'RIFF';
       Move(text[0], MT.WaveHdr.RIFF[0], 4);
       MT.WaveHdr.FileSize := MT.FSize + 44;
       text := 'WAVE';
       Move(text[0], MT.WaveHdr.RIFFType[0], 4);
       text := 'fmt ';
       Move(text[0], MT.WaveHdr.FmtChunkId[0], 4);
       MT.WaveHdr.FmtChunkSize := 16;
       MT.WaveHdr.FormatTag := 1;
       MT.WaveHdr.Channels := MT.FChan;
       MT.WaveHdr.SampleRate := MT.FSR;
       MT.WaveHdr.BitsPerSample := 16;
       MT.WaveHdr.BlockAlign := 2*MT.FChan;
       MT.WaveHdr.BytesPerSecond := MT.FSR * MT.WaveHdr.BlockAlign;
       text := 'data';
       Move(text[0], MT.WaveHdr.DataChunkId[0], 4);
       MT.WaveHdr.DataSize := MT.FSize;
       if MT.FOutputStream is TMemoryStream then
       begin
         MT.FOutputStream.Size :=MT.FSize + 44;
         MT.FOutputStream.Seek(0, soFromBeginning);
       end;
       MT.FOutputStream.Write(MT.WaveHdr, 44);
       MT.HasFirstFrame := True;
     end;
     if pcm.channels = 2 then
     begin
       for i := 0 to pcm.length -1 do
       begin
         if pcm.samples[0][i] >= MAD_F_ONE then
         pcm.samples[0][i] := MAD_F_ONE - 1;
         if pcm.samples[0][i] < -MAD_F_ONE then
         pcm.samples[0][i] := -MAD_F_ONE;
         pcm.samples[0][i] := pcm.samples[0][i] shr (MAD_F_FRACBITS + 1 - 16);
         outsamples[i shl 1] := pcm.samples[0][i];
         if pcm.samples[1][i] >= MAD_F_ONE then
         pcm.samples[1][i] := MAD_F_ONE - 1;
         if pcm.samples[1][i] < -MAD_F_ONE then
         pcm.samples[1][i] := -MAD_F_ONE;
         pcm.samples[1][i] := pcm.samples[1][i] shr (MAD_F_FRACBITS + 1 - 16);
         outsamples[(i shl 1)+1] := pcm.samples[1][i];
       end;
       MT.FOutputStream.Write(outsamples[0], pcm.length*4);
     end else
     begin
       for i := 0 to pcm.length -1 do
       begin
         if pcm.samples[0][i] >= MAD_F_ONE then
         pcm.samples[0][i] := MAD_F_ONE - 1;
         if pcm.samples[0][i] < -MAD_F_ONE then
         pcm.samples[0][i] := -MAD_F_ONE;
         pcm.samples[0][i] := pcm.samples[0][i] shr (MAD_F_FRACBITS + 1 - 16);
         outsamples[i] := pcm.samples[0][i];
       end;
       MT.FOutputStream.Write(outsamples[0], pcm.length*2);
     end;
     if MT.FSize <> 0 then
     begin
       CProgr := Round(MT.FOutputStream.Position/MT.FSize*100);
       if MT.Progr <> CProgr then
       begin
         MT.Progr := CProgr;
         if Assigned(MT.FMADProgress) then
         MT.FMADProgress(MT.Owner);
       end;
     end;
     if not MT.Terminated then Result := MAD_FLOW_CONTINUE
     else Result := MAD_FLOW_STOP;
   end;

   function ErrorFunc(CData : Pointer; Stream : p_mad_stream; Frame : p_mad_frame) : Integer; cdecl;
   begin
     Result := MAD_FLOW_CONTINUE;
   end;

   constructor TMADThread.Create;
   begin
     inherited Create(True);
     Owner := AOwner;
     FInputStream := InputStream;
     FOutputStream := OutputStream;
     FreeOnTerminate := False;
   end;

   destructor TMADThread.Destroy;
   begin
     if not _Free then
     begin
       Terminate;
       {$IFDEF WIN32}
       while not _Free do;
       {$ENDIF}
     end;
     inherited Destroy;
   end;

   procedure TMADThread.Execute;
   begin
     try
       GetMem(Data, FInputStream.Size);
       FInputStream.Read(Data[0], FInputStream.Size);
       mad_decoder_init(@FDecoder, Self, InputFunc, nil, nil, OutputFunc, ErrorFunc, nil);
       mad_decoder_run(@FDecoder, MAD_DECODER_MODE_SYNC);
       mad_decoder_finish(@FDecoder);
       FreeMem(Data);
       WhenDone;
       if Assigned(FMADDone) then FMADDone(Owner, FValid);
       _Free := True;
     except
       FreeMem(Data);
       WhenDone;
       _Free := True;
     end;
   end;

end.
